### ISSUE TITLE
Dataset and Dataloader - Batch-size and rounding-related parameters

### ISSUE THREAD

**Issue by fKunstner** [fKunstner/optexp#10]
What would be a reasonable way to handle the parameters related to the batch size and order of minibatches?
```
batch_size
micro_batch_size
shuffle
drop_last
sampler
```

Complete list of requirements that is not feasible to hit:
- Make it easy to run full batch experiments
- Make it easy to compare sampling IID and shuffling
- Stay close to Pytorch behavior defaults
- Reproducible behavior across micro batch size so its selection can be automated
- Fit all the above with default parameters

## The `drop_last` problem

Suppose we have `N = number_of_samples`, `B = batch_size`, and for simplicity `batch_size == micro_batch_size`. 

The thing we want to avoid is having a step with a sample size smaller than `B`.

I can see at least two ways to handle `drop_last`; 
- 1. Use only a subset of the data at each epoch, miss some samples at each epoch (Pytorch default).
- 2. Truncate the dataset at the start and use all the data at each epoch.

Both miss on `N mod B < B` samples, but (1) misses them randomly at every while (2) doesn't miss any but on a reduced dataset.

### Additional difficulty 1: Interaction with multiprocessing
Using multiprocessing, `drop_last` is applied per dataloader on each worker node.
With `W` workers, we will miss `N mod B W < BW` samples

### Additional difficulty 2: Interaction with micro batch size /!\
The dataloader will iterate over blocks of `micro_batch_size`, not `batch_size`.
But if the dataloader contains less than `B` samples less, we still shouldn't take a step, because it would be incomplete?
This cannot be implemented with the default pytorch dataloader and `drop_last`, which will drop when the remaining samples is less than `microB`.

With multiprocessingg, we should stop when the local dataloader contains less than `B/W` samples.

_But_, if we leave `drop_last=False`, then we don't actually _need_ to have that `B mod microB == 0`, 
because we can accomodate any batch size. Although we might have to randomly drop samples..?
(Maybe too much in the weeds) 

### Additional difficulty 3: Consistency of the "problem" across seeds /!\

- If we do 1. 
  - and disable shuffling: we never get to optimize a small subset of the dataset.
  - and don't disable shuffling: we don't run in full batch but in almost full batch.
- If we do 2. 
  - and use a different seed for the dataset: we have different datasets will lead to different problems every time.
  - and use the same for the dataset: we have the same dataset every time, but we have to make sure the validation dataloader is consistent with that.

### Additional difficulty 4: Difference between training and evaluation /!\
We don't need to `drop_last` on the evaluation datasets if we're only using them to go through 

## Proposed solution 

?

## Sanity check:
- How does it work if we want to run in full batch?
  - Ideally we would just have to set `B = N`, instead of having to set a separate `full_batch` mode
    - How do we know what `N` is? 
      - What if it changes as a function of the batch size?
        - It doesn't (?), in the 
    - What if `N` is a prime number and there is no way to set a reasonable micro batch size?
